tools/cluster_mgr.py

```diff
@@ -153,11 +149,180 @@ def main():
+        }
+
+    def build_slots(slot_list):
+        slots = []
+        for i in range(0, len(slot_list), 2):
+            slots.append(
+                {
+                    "start": slot_list[i],
+                    "end": slot_list[i+1]
+                }
+            )
+        return slots
+
+    client = redis.Redis(decode_responses=True,
+                         host="localhost", port=args.first_port)
+    existing = client.execute_command("cluster", "shards")
+    config = []
+    for shard_list in existing:
+        shard = list_to_dict(shard_list)
+        config.append({
+            "slot_ranges": build_slots(shard["slots"]),
+            "master": build_node(shard["nodes"][0]),
+            "replicas": [build_node(replica) for replica in shard["nodes"][1::]]
+        })
+    return config
+
+
+def move(args):
+    config = build_config_from_existing(args)
+
+    new_owner = None
+    for shard in config:
+        if shard["master"]["port"] == args.target_port:
+            new_owner = shard
+            break
+    else:
+        print(
+            f"Can't find master with port {args.target_port} (hint: use flag --target_port).")
+        exit(-1)
+
+    def remove_slot(slot, from_range, from_shard):
+        if from_range["start"] == slot:
+            from_range["start"] += 1
+            if from_range["start"] > from_range["end"]:
+                from_shard["slot_ranges"].remove(from_range)
+        elif from_range["end"] == slot:
+            from_range["end"] -= 1
+            if from_range["start"] > from_range["end"]:
+                from_shard["slot_ranges"].remove(from_range)
+        else:
+            assert slot > from_range["start"] and slot < from_range["end"], f'{slot} {from_range["start"]} {from_range["end"]}'
+            from_shard["slot_ranges"].append(
+                {"start": slot + 1, "end": from_range["end"]})
+            from_range["end"] = slot - 1
+
+    def add_slot(slot, to_shard):
+        for slot_range in to_shard["slot_ranges"]:
+            if slot == slot_range["start"] - 1:
+                slot_range["start"] -= 1
+                return
+            if slot == slot_range["end"] + 1:
+                slot_range["end"] += 1
+                return
+        to_shard["slot_ranges"].append({"start": slot, "end": slot})
+
+    def find_slot(slot, config):
+        for shard in config:
+            if shard == new_owner:
+                continue
+            for slot_range in shard["slot_ranges"]:
+                if slot >= slot_range["start"] and slot <= slot_range["end"]:
+                    return shard, slot_range
+        return None, None
+
+    def pack(slot_ranges):
+        for slot_range in slot_ranges:
+            for other_slot_range in slot_ranges:
+                if slot_range["end"] + 1 == other_slot_range["start"]:
+                    slot_range["end"] = other_slot_range["end"]
+                    slot_ranges.remove(other_slot_range)
+                    break
```

Comment:
1. Not sure what happens when you remove an element while iterating (i.e. other_slot_range is before slot_range and the array moves). Or is it sorted? In that case you don't need the double loop
2. I assume you have only a few ranges usually, so creating a new object would solve all issues

Reply:
Good feedback! Thanks!
Like you suggested, the new version of this function creates a new range, and iterates until no changes are detected.
I also sorted the ranges to ease things up, even though it doesn't have to be sorted.